### PR TITLE
Migrated db utilities from a10-openstack

### PR DESCRIPTION
**Pull request by mdurrant-b3**

### CODE REVIEW

File: a10_openstack_lib/dbmigration/__init__.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2012 New Dream Network, LLC (DreamHost)
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/__init__.py]: Just to double check, is this a straight copy or did we modify it?

File: a10_openstack_lib/dbmigration/alembic.ini [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+class = StreamHandler
+args = (sys.stderr,)
+level = NOTSET
+formatter = generic
+
+[formatter_generic]
+format = %(levelname)-5.5s [%(name)s] %(message)s
+datefmt = %H:%M:%S
+
+
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic.ini]: Extraneous whitespace.

File: a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+def upgrade_l3():
+    op.create_table(
+        'routers',
+        sa.Column('tenant_id', sa.String(length=255), nullable=True),
+        sa.Column('id', sa.String(length=36), nullable=False),
+        sa.Column('name', sa.String(length=255), nullable=True),
+        sa.Column('status', sa.String(length=16), nullable=True),
+        sa.Column('admin_state_up', sa.Boolean(), nullable=True),
+        sa.Column('gw_port_id', sa.String(length=36), nullable=True),
+        sa.ForeignKeyConstraint(['gw_port_id'], ['ports.id'], ),
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py]: Extra `, ` before `)` unneeded?

File: a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        sa.Column('id', sa.String(length=36), nullable=False),
+        sa.Column('floating_ip_address', sa.String(length=64), nullable=False),
+        sa.Column('floating_network_id', sa.String(length=36), nullable=False),
+        sa.Column('floating_port_id', sa.String(length=36), nullable=False),
+        sa.Column('fixed_port_id', sa.String(length=36), nullable=True),
+        sa.Column('fixed_ip_address', sa.String(length=64), nullable=True),
+        sa.Column('router_id', sa.String(length=36), nullable=True),
+        sa.ForeignKeyConstraint(['fixed_port_id'], ['ports.id'], ),
+        sa.ForeignKeyConstraint(['floating_port_id'], ['ports.id'], ),
+        sa.ForeignKeyConstraint(['router_id'], ['routers.id'], ),
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py]: Extra `, ` before `)` unneeded here and other 2 lines above?

File: a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        sa.PrimaryKeyConstraint('id')
+    )
+
+
+def upgrade_quota(options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py]: Unused param `options`.

File: a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        sa.PrimaryKeyConstraint('id')
+    )
+
+
+def downgrade_l3():
+    for table in ('floatingips', 'routers', 'externalnetworks'):
+        op.drop_table(table)
+
+
+def downgrade_quota(options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/common_ext_ops.py]: Unused param `options`.

File: a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# Copyright (C) ${create_date.year} A10 Networks Inc. All rights reserved.
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako]: Should this be `${create_date.year},`?

File: a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+]
+
+from alembic import op
+import sqlalchemy as sa
+${imports if imports else ""}
+
+from neutron.db import migration
+
+
+def upgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    if not migration.should_run(active_plugins, migration_for_plugins):
+        return
+
+    ${upgrades if upgrades else "pass"}
+
+
+def downgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+]
+
+from alembic import op
+import sqlalchemy as sa
+
+
+from neutron.db import migration
+
+
+def upgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+                    sa.Column('devices_id', sa.Integer),  # TODO(mdurrant) Should this be device ID?
+                    sa.Column('host', sa.String(45)),  # TODO(mdurrant) Max hostname length is 255
+                    sa.Column('is_valid', sa.Integer),
+                    sa.PrimaryKeyConstraint('id'),
+                    mysql_engine='InnoDB'
+                    )
+    op.create_index('vrrp_id_UNIQUE', 'a10_zone', ['vrid', 'host', 'name'], unique=True)
+
+
+def downgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+]
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import mysql
+
+from neutron.db import migration
+
+
+def upgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    sa.ForeignKeyConstraint(['certificate_id'], ['certificates.id'], ),
+    sa.ForeignKeyConstraint(['vip_id'], ['vips.id'], ),
+    sa.PrimaryKeyConstraint('id'),
+    mysql_engine='InnoDB'
+    )
+
+    ### end Alembic commands ###
+
+
+def downgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py]: Unused param `options`?

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/a10base.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+]
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import mysql
+
+from neutron.db import migration
+
+
+def upgrade(active_plugins=None, options=None):
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/a10base.py]: Unused param `options`?

File: a10_openstack_lib/dbmigration/cli.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    config = alembic_config.Config(
+        os.path.join(os.path.dirname(__file__), 'alembic.ini')
+    )
+    config.set_main_option('script_location',
+                           SCRIPT_LOCATION)
+    # attach the Neutron conf to the Alembic conf
+    config.neutron_config = CONF
+
+    CONF()
+    #TODO(gongysh) enable logging
```
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/cli.py]: Space after `#`

File: a10_openstack_lib/dbmigration/__init__.py (hunk 1 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/__init__.py]: Straight copy.

File: a10_openstack_lib/dbmigration/alembic.ini (hunk 2 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic.ini]: Does PEP apply to INI files?

File: a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako (hunk 8 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako]: This is a template - options could be used by generated files.  I'd have to look at the cli.py to see if it adds stuff to this.

File: a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako (hunk 9 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/script.py.mako]: This is a template - options could be used by generated files. I'd have to look at the cli.py to see if it adds stuff to this.

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py (hunk 10 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py]: Auto-gen'd code.

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py (hunk 11 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/19703d155d7d_a10_vxlan.py]: Auto-gen'd code.

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py (hunk 12 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py]: Auto-gen'd code.

File: a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py (hunk 13 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/alembic_migrations/versions/46b0cf8fe805_certificates_and_bindings.py]: Auto-gen'd code.

File: a10_openstack_lib/dbmigration/cli.py (hunk 15 above)
Comment (mdurrant-b3) [inline a10_openstack_lib/dbmigration/cli.py]: OK you got me there, though it is someone else's comment. :)

File: a10_openstack_lib/dbmigration/alembic.ini (hunk 2 above)
Comment (0xtobit) [inline a10_openstack_lib/dbmigration/alembic.ini]: No

**Pull request merged**